lib.py

```diff
@@ -1,167 +1,284 @@
 import asyncio
 from datetime import datetime, timedelta, timezone
 import functools
+import json
 import logging
-import traceback
+import pathlib
+import shutil
+from sqlite3 import Connection, Cursor
+import statistics as stats
+import tempfile
 import os
 from zoneinfo import ZoneInfo
 
+import discord
 import docker
 from database import Database
 from messages import SubmitMessage
 
+from config import settings
+
 doc = docker.from_env()
 
 
 logger = logging.getLogger(__name__)
 
 
-async def build_image(msg, solution):
-    logger.info("Building for %s", msg.author.name)
-    status = await msg.reply("Building...")
-    with open("runner/src/code.rs", "wb+") as f:
-        f.write(solution)
+async def benchmark(submit_msg: SubmitMessage):
+    msg, day, code, part = submit_msg.msg, submit_msg.day, submit_msg.code, submit_msg.part
+    op_name, op_id = msg.author.name, msg.author.id
+
+    try:
+        db = Database().get()
+        cur = db.cursor()
+        results = []
+        with tempfile.TemporaryDirectory(suffix=f"-ferris-elf-{op_id}") as tmpdir:
+            populate_tmp_dir(tmpdir, code)
+            if not await build_code(op_name, op_id, tmpdir):
+                # This reply is not good UX, but it's better than silence.
+                await msg.reply("Build failed.")
+                return
+            answers_map = load_answers(cur, day, part)
+            for in_file in get_input_files(day):
+                logger.info("Processing file: %s", in_file)
+                load_input(tmpdir, day, in_file)
+                result_lst = await run_code(op_name, op_id, tmpdir, in_file)
+                result = process_run_result(in_file, answers_map, result_lst)
+                results.append(result)
+            
+            if len(results) > 0:
+                save_results(cur, op_id, day, part, code, results)
+                db.commit()
+            
+        verified_results = [r for r in results if r.get("verified", False)]
+        if len(verified_results) > 0:
+            median = stats.mean([r["median"] for r in verified_results])
```

Comment:
- Criterion itself prints out the values in "[typical](https://github.com/bheisler/criterion.rs/blob/b913e232edd98780961ecfbae836ec77ede49259/src/report.rs#L555)" instead of mean or median. According to [this](https://bheisler.github.io/criterion.rs/book/cargo_criterion/external_tools.html) this is equal to slope (not sure what this is) or mean if available.

Mostly a note that we may want to report the typical value or investigate later. 

- We should maybe also write in this message that the median value is being used for ranking.

Reply:
I don't have evidence for this, but I'm guessing that criterion is doing a line-fitting. Especially since the raw sample count increases linearly with each sample.

I like the idea of noting in the constants what's going on, but if I take that work on myself, what I'd probably do is spend a couple hours figuring out what Criterion prints on the command line and HTML reports. If the CSV reports, command line output, and HTML reports all use "typical" (seems likely), then it probably makes sense to change the code in this PR to do the same. Simpler for users, simpler for the docs.